### PR TITLE
fix: hang in Deno.serveHttp()

### PR DESCRIPTION
**Pull request by bartlomieju**
Waiting on next request in `Deno.serveHttp()` API hanged
when responses were using `ReadableStream`. This was caused
by `op_http_request_next` op that was never woken after
response was fully written. This commit adds `waker` field to 
`DenoService` which is called after response is finished.

Fixes https://github.com/denoland/deno/issues/10870

### CODE REVIEW

File: cli/tests/unit/http_test.ts [hunk 1]
```diff
@@ -373,3 +373,45 @@ unitTest(
     await delay(300);
   },
 );
+
+unitTest(
+  { perms: { net: true } },
+  async function httpServerHang() {
```
Comment (bartlomieju) [inline cli/tests/unit/http_test.ts:379]: This test is actually useless, it also passes on `main`

Comment (bartlomieju) [reply cli/tests/unit/http_test.ts:383]: Updated the test and made sure it hangs on `main`

File: cli/tests/unit/http_test.ts [hunk 2]
```diff
@@ -373,3 +377,74 @@ unitTest(
...
+            }
+          }
+        })();
+      }
+    })();
+
+    const clientConn = await Deno.connect({ port: 4501 });
+    const encoder = new TextEncoder();
+
+    async function requestAndReadResponse(conn: Deno.Conn): Promise<string> {
```
Comment (ry) [inline cli/tests/unit/http_test.ts:414]: Maybe move this utility function to the top-level? Might be useful in other tests in the future.

File: runtime/ops/http.rs [hunk 3]
```diff
@@ -409,6 +410,9 @@ async fn op_http_response(
   })
   .await?;
 
+  if maybe_response_body_rid.is_none() {
+    conn_resource.deno_service.waker.wake();
+  }
```
Comment (ry) [inline runtime/ops/http.rs:415]: I mean, sure, whatever makes the tests pass.

But I wonder if the waker.wake() could be done more centrally inside `conn_resource.poll` but only when it's about to return pending...

File: cli/tests/unit/http_test.ts (hunk 2 above)
Comment (bartlomieju) [reply cli/tests/unit/http_test.ts:414]: Sure

File: runtime/ops/http.rs (hunk 3 above)
Comment (bartlomieju) [reply runtime/ops/http.rs:415]: > But I wonder if the waker.wake() could be done more centrally inside conn_resource.poll but only when it's about to return pending...

But we don't want that inside `conn_resource.poll` - if we did that in `conn_resource.poll` then we'd be waking the future of `op_http_request_next` while polling that op. Actually we want to wake `op_http_request_next` when a response has finished writing which can happen either in `op_http_response` or `op_http_response_close `.

File: runtime/ops/http.rs [hunk 4]
```diff
@@ -160,15 +161,16 @@ async fn op_http_request_next(
   let cancel = RcRef::map(conn_resource.clone(), |r| &r.cancel);
 
   poll_fn(|cx| {
+    conn_resource.deno_service.waker.register(cx.waker());
     let connection_closed = match conn_resource.poll(cx) {
       Poll::Pending => false,
       Poll::Ready(Ok(())) => {
-        // close ConnResource
-        state
+        // try to close ConnResource, but don't unwrap as it might
```
Comment (piscisaureus) [inline runtime/ops/http.rs:168]: If the conn could already be closed, doesn't this run the risk of closing the wrong resource?

Comment (bartlomieju) [reply runtime/ops/http.rs:168]: Not really - `ConnResources` might have been taken by one of the other ops and we don't use `rid`s so there's no risk that you close random resource by mistake.

**Comment by bartlomieju** [comment]
> I suspect the problem is really here:
> 
> https://github.com/denoland/deno/blob/f7c59684fb8b00afce8ee89a50aed066493650a5/runtime/ops/http.rs#L84
> 
> Remember the rule of futures club: if you return Pending you have to arrange a wake-up. So I think you should capture the context waker at right there, and wake that waker when self.inner is cleared.

Sure that makes sense to me, but the thing is - I chased all the polls to `ConnResource` in other ops and `DenoService::call` is never called in any of the cases so effectively `self.inner` would never wake up this task.

> (A message from our sponsor: AsyncRefCell takes care of all that. Great discounts for new customers.)

Not sure what you mean by here, what should be wrapped in `AsyncRefCell`?

**Review by lucacasonato — COMMENTED** [review]
I have a hunch that this still will hang if you delay responding to the incoming request.

A test case would be to send 10 requests at once on a H2 connection, and have it take 5 seconds for them to send back headers. Are they processed all at once, or serially? I can try write this test case and open a PR for it.

**Review by ry — APPROVED** [review]
LGTM - nice work @bartlomieju !

@lucacasonato A test case like that sounds very useful.

**Comment by piscisaureus** [comment]
> Sure that makes sense to me, but the thing is - I chased all the polls to `ConnResource` in other ops and `DenoService::call` is never called in any of the cases so effectively `self.inner` would never wake up this task.

I don't know what this is about.

Suggestion: 
* before returning `Poll::Pending` [here](https://github.com/denoland/deno/blob/f7c59684fb8b00afce8ee89a50aed066493650a5/runtime/ops/http.rs#L84), capture the waker (`inner.waker.register(cx.waker())`).
* [here](https://github.com/denoland/deno/blob/f7c59684fb8b00afce8ee89a50aed066493650a5/runtime/ops/http.rs#L196), wake the waker (`conn_resource.deno_service.inner.waker.wake()`)

**Pull request merged**